project_cost_report/report.py

```diff
@@ -0,0 +1,436 @@
+        :rtype: dict
+        """
+        domain = self._get_time_analytic_line_domain(project)
+        lines = self.env['account.analytic.line'].search(domain)
+        grouped_lines = _group_analytic_lines(lines, lambda l: l.task_id.task_type_id)
+        sorted_categories = sorted(grouped_lines.keys(), key=lambda c: c.name or "")
+        unfolded_categories = report_context.get('unfolded_categories') or {}
+        unfolded_time_categories = unfolded_categories.get('time') or []
+        return [
+            TimeCategory(
+                id_=c.id,
+                name=c.name or self._get_empty_task_type_label(),
+                lines=grouped_lines.get(c),
+                folded=c.id not in unfolded_time_categories
+            ) for c in sorted_categories
+        ]
+
+    def _get_time_total(self, project):
+        """Get the total amount for the TIME section.
+
+        :param project: the project.project record
+        :rtype: float
+        """
+        domain = self._get_time_analytic_line_domain(project)
+        result = self.env['account.analytic.line'].read_group(
+            domain=domain, fields=['account_id', 'amount'], groupby='account_id')
+        amount = result[0]['amount'] if result else 0
+        return float_round(-amount, 2)
+
+    def _get_time_total_hours(self, project):
+        """Get the total hours (units) for the TIME section.
+
+        :param project: the project.project record
+        :rtype: float
+        """
+        domain = self._get_time_analytic_line_domain(project)
+        result = self.env['account.analytic.line'].read_group(
+            domain=domain, fields=['account_id', 'unit_amount'], groupby='account_id')
+        amount = result[0]['unit_amount'] if result else 0
+        return float_round(amount, 2)
+
+    def _get_rendering_variables(self, project, report_context):
+        """Add the variables related to the TIME section."""
+        res = super()._get_rendering_variables(project, report_context)
+        res.update({
+            'time_categories': self._get_time_categories(project, report_context),
+            'time_total': self._get_time_total(project),
+            'time_total_hours': self._get_time_total_hours(project),
+        })
+        return res
+
+    @api.model
+    def get_foldable_categories(self, project_id):
+        """Add time categories to foldable categories."""
+        res = super().get_foldable_categories(project_id)
+        project = self.env['project.project'].browse(project_id)
+        domain = self._get_time_analytic_line_domain(project)
+        analytic_lines = self.env['account.analytic.line'].search(domain)
+        res['time'] = analytic_lines.mapped('task_id.task_type_id.id')
+        res['time'].append(False)  # Empty task type category (Labour)
+        return res
+
+
+class ProjectCostReportWithOutsourcing(models.TransientModel):
+
+    _inherit = 'project.cost.report'
+
+    def _get_empty_outsourcing_category_label(self):
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Method could be a function](https://app.codacy.com/app/numigi/odoo-project-addons/pullRequest?prid=3675949)